apiserver/common/tools.go

```diff
@@ -256,6 +251,23 @@ func (f *ToolsFinder) findTools(args params.FindToolsParams) (coretools.List, er
 	if err != nil {
 		return nil, err
 	}
+
+	// This handles clients and agents that may be attempting to find tools in
+	// the context of series instead of OS type.
+	// If we get a request by series we ensure that any matched OS tools are
+	// converted to the requested series.
+	// Conversely, if we get a request by OS type, matching series tools are
+	// converted to match the OS.
+	// TODO: Remove this block and the called methods for Juju 3/4.
+	if args.Number.Major == 2 && args.Number.Minor <= 8 && (args.OSType != "" || args.Series != "") {
```

Comment:
I really wish this could be a facade version bump for 2.9 rather than a version number check.

Reply:
This is potentially preamble to such a change. The issue is that `ToolsFinder` is used both as an embedded facade, and as a general tools finding class - even in `ToolsGetter` below, which means we'd have to give that the same treatment.

I'm with you on preferring the facade versioning mechanism. I will endeavour to follow up on it.